File: src/main/java/org/example/expert/config/security/loginuser/LoginUserService.java
package org.example.expert.config.security.loginuser;

import lombok.RequiredArgsConstructor;
import org.example.expert.domain.user.repository.UserRepository;
import org.springframework.security.authentication.InternalAuthenticationServiceException;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class LoginUserService implements UserDetailsService {

    private final UserRepository userRepository;
    @Override
    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
        return userRepository.findByEmail(email)
                .map(LoginUser::new)
                .orElseThrow(() -> new InternalAuthenticationServiceException("인증 실패"));
    }
}

File: src/main/java/org/example/expert/config/security/loginuser/LoginUserService.kt
package org.example.expert.config.security.loginuser

import org.example.expert.domain.user.UserRepository
import org.springframework.security.authentication.InternalAuthenticationServiceException
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class LoginUserService(
    private val userRepository: UserRepository
): UserDetailsService {
    override fun loadUserByUsername(email: String): UserDetails {
        return userRepository.findByEmail(email)
            ?.let { LoginUser(it) }
            ?: throw InternalAuthenticationServiceException("인증 실패")
    }
}